pkg/metrics/registry.go

```diff
@@ -0,0 +1,135 @@
+// Dynamic histograms
+// -----------------------------------------------------------------------------
+
+func (r *MetricRegistry) GetOrCreateDynamicHistogram(name, help string) *DynamicHistogram {
+	r.mu.RLock()
+	if h, ok := r.histograms[name]; ok {
+		r.mu.RUnlock()
+		return h
+	}
+	r.mu.RUnlock()
+
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	if h, ok := r.histograms[name]; ok {
+		return h
+	}
+
+	h := NewDynamicHistogram(name, help)
+	prometheus.MustRegister(h)
+	r.histograms[name] = h
+	return h
+}
+
+// -----------------------------------------------------------------------------
+// Authconfig metrics
+// -----------------------------------------------------------------------------
+
+func (r *MetricRegistry) GetAuthServerAuthConfigTotalMetric() *DynamicCounter {
+	return r.GetOrCreateDynamicCounter(
+		"auth_server_authconfig_total",
+		"Total number of authconfigs enforced by the auth server, partitioned by authconfig.",
+	)
+}
+
+func (r *MetricRegistry) GetAuthServerAuthConfigResponseStatusMetric() *DynamicCounter {
+	return r.GetOrCreateDynamicCounter(
+		"auth_server_authconfig_response_status",
+		"Response status of authconfigs sent by the auth server, partitioned by authconfig.",
+	)
+}
+
+func (r *MetricRegistry) GetAuthServerAuthConfigDurationMetric() *DynamicHistogram {
+	return r.GetOrCreateDynamicHistogram(
+		"auth_server_authconfig_duration_seconds",
+		"Response latency of authconfig enforced by the auth server (in seconds).",
+	)
+}
+
+// -----------------------------------------------------------------------------
+// Evaluator metrics
+// -----------------------------------------------------------------------------
+
+func (r *MetricRegistry) GetAuthServerEvaluatorTotalMetric() *DynamicCounter {
+	return r.GetOrCreateDynamicCounter(
+		"auth_server_evaluator_total",
+		"Total number of evaluations of individual authconfig rule performed by the auth server.",
+	)
+}
+
+func (r *MetricRegistry) GetAuthServerEvaluatorCancelledMetric() *DynamicCounter {
+	return r.GetOrCreateDynamicCounter(
+		"auth_server_evaluator_cancelled",
+		"Number of evaluations of individual authconfig rule cancelled by the auth server.",
+	)
+}
+
+func (r *MetricRegistry) GetAuthServerEvaluatorIgnoredMetric() *DynamicCounter {
+	return r.GetOrCreateDynamicCounter(
+		"auth_server_evaluator_ignored",
+		"Number of evaluations of individual authconfig rule ignored by the auth server.",
+	)
+}
+
+func (r *MetricRegistry) GetAuthServerEvaluatorDeniedMetric() *DynamicCounter {
+	return r.GetOrCreateDynamicCounter(
+		"auth_server_evaluator_denied",
+		"Number of denials from individual authconfig rule evaluated by the auth server.",
+	)
+}
+
+func (r *MetricRegistry) GetAuthServerEvaluatorDurationMetric() *DynamicHistogram {
+	return r.GetOrCreateDynamicHistogram(
+		"auth_server_evaluator_duration_seconds",
+		"Response latency of individual authconfig rule evaluated by the auth server (in seconds).",
+	)
+}
+
+var Registry = NewMetricRegistry()
```

Comment:
We used to define AuthPipeline-related metrics in the `service` package, for example. This made more sense to me than having some metric definitions widespread and others contained within the packages they are used.

Reply:
Yeah, fair. I can move this to `service` package